bitswap: Bitswap now sends multiple blocks per message

taylormike opened the pull request:
Updated PeerRequestTask to hold multiple wantlist.Entry(s). This allows Bitswap to send multiple blocks in bulk per a Peer's request. Also, added a metric for how many blocks to put in a given message. Currently: 512 * 1024 bytes. 
Fixes [#4378](https://github.com/ipfs/go-ipfs/issues/4378)
@whyrusleeping @Stebalien

whyrusleeping commented:
Hey @taylormike, sorry for the wait. New repos mean i wasnt watching for notifications.

whyrusleeping commented on decision/engine.go:
```diff
@@ -231,6 +240,8 @@ func (e *Engine) MessageReceived(p peer.ID, m bsmsg.BitSwapMessage) error {
 		l.wantList = wl.New()
 	}
 
+	blockSize, msgSize := 0, 0
```
cleaner to write:
```
var blockSize, msgSize int
```

taylormike replied:
@whyrusleeping I agree.
This is done. I pushed out these changes and squashed into a single commit.

whyrusleeping commented on decision/engine.go:
```diff
@@ -240,12 +251,26 @@ func (e *Engine) MessageReceived(p peer.ID, m bsmsg.BitSwapMessage) error {
 			if exists, err := e.bs.Has(entry.Cid); err == nil && exists {
-				e.peerRequestQueue.Push(entry.Entry, p)
 				newWorkExists = true
+				if blockSize, _ = e.bs.GetSize(entry.Cid); blockSize == -1 {
```
I wouldnt throw away an error. If this actually errors, something seems pretty wrong. It's worthy of logging IMO

schomatis replied:
There was a previous mention from @Stebalien that `GetSize` wouldn't fail, but maybe we could add a comment there explaining it, since yes, that silent error seems suspicious.

> This really shouldn't be necessary. GetSize should always work (even if it isn't cached).

Stebalien replied:
Sorry, that statement was a bit confusing. I meant that if the `Get` worked, the `GetSize` should have worked. That is, we can just call:

```go
size, err := e.bs.GetSize(entry.Cid)
if err != nil {
    /* we don't have it, *may* be an error, check if it's bs.ErrNotFound */
} else {
    // we have the block
}
```

We don't need to call `GetSize`, then `Get`, then `GetSize`.

taylormike replied:
@whyrusleeping, @schomatis, @schomatis  I agree. 
I removed the 'GetSize, then Get, then GetSize sequence' and I'm now logging the error. 

This is done. I pushed out these changes and squashed into a single commit.

taylormike commented on decision/engine.go:
```diff
@@ -240,12 +251,26 @@ func (e *Engine) MessageReceived(p peer.ID, m bsmsg.BitSwapMessage) error {
+					activeEntries = []*wl.Entry{}
+					msgSize = 0
+				}
+				activeEntries = append(activeEntries, entry.Entry)
```
This is intentional.
That being said, I can rewrite it as such:
```Go
  if msgSize + blockSize < maxMessageSize {
       activeEntries = append(activeEntries, entry.Entry)
       msgSize += blockSize 
       } else {
	 e.peerRequestQueue.Push(p, activeEntries...)
	 activeEntries = []*wl.Entry{}
	 msgSize = 0
        }
```

taylormike commented on decision/peer_request_queue.go:
```diff
@@ -46,7 +46,7 @@ type prq struct {
 
 // Push currently adds a new peerRequestTask to the end of the list
-func (tl *prq) Push(entry *wantlist.Entry, to peer.ID) {
+func (tl *prq) Push(entries []*wantlist.Entry, to peer.ID) {
```
@Stebalien I agree.
This is done. I pushed out these changes and squashed into a single commit.

taylormike commented:
I pushed out the most recent code review feedback  changes and squashed into a single commit.
I also updated bench_test, peer_request_queue_test, and engine_test accordingly.

Stebalien commented on decision/engine.go:
```diff
@@ -240,12 +251,24 @@ func (e *Engine) MessageReceived(p peer.ID, m bsmsg.BitSwapMessage) error {
 			if exists, err := e.bs.Has(entry.Cid); err == nil && exists {
-				e.peerRequestQueue.Push(entry.Entry, p)
 				newWorkExists = true
+				if blockSize, err = e.bs.GetSize(entry.Cid); err != nil {
```
Actually, we should probably just use `GetSize` instead of `Has`. That way, we only make one request.

Stebalien replied:
(that is, we can replace the `Has` check in the condition)

taylormike replied:
@Stebalien 
I agree, I pushed out this change and squashed into a single commit.

Stebalien commented on decision/peer_request_queue.go:
```diff
@@ -195,7 +222,16 @@ func (t *peerRequestTask) SetIndex(i int) {
-func taskKey(p peer.ID, k *cid.Cid) string {
+func taskKey(p peer.ID, entries []*wantlist.Entry) string {
+	key := string(p)
+	for _, entry := range entries {
```
This is a bit unfortunate as it could grow quite large (*very* large). I wonder if we can switch to a different method for computing task keys.

taylormike replied:
@Stebalien  Sorry for the delayed response. 
This is done. I pushed out this change and squashed into a single commit.

Stebalien commented:
I've tested this against go-ipfs and it doesn't appear to break anything.

Stebalien commented:
It looks like the checkHandledInOrder test is broken.

taylormike commented:
@Stebalien I'm looking into the checkHandledInOrder test failure.

taylormike commented:
@Stebalien  Sorry for the delayed response. 
The checkHandledInOrder test failure is now resolved. I pushed out this change and squashed into a single commit.

Stebalien commented:
Running the go-ipfs sharness tests now.

Stebalien commented:
Looks good!
